File: examples/csharp/avalonia/NotesShowcase/ViewModels/NoteVM.cs
using System.Windows.Input;
using VMx.Builders;
using VMx.Capabilities;
using VMx.Commands;
using VMx.Components;
using VMx.Messages;
using VMx.Notifications;
using VMx.Services;
using NotesShowcase.Models;

namespace NotesShowcase.ViewModels;

/// <summary>
/// Leaf VM for a single note.
///
/// Capabilities (plan §3.a.6, scenario §6.2):
///   <see cref="ISelectable"/>, <see cref="IClosable"/>,
///   <see cref="IDeletable{NoteVM}"/>, <see cref="ISavable{NoteVM}"/>,
///   <see cref="IReconstructable"/>.
///
/// <see cref="CloseCommand"/> invokes a host-supplied close callback (the host
/// wires it to <c>NotesViewVM.Current = null</c> so the form clears). This
/// avoids back-references from the leaf to its container.
/// </summary>
public sealed class NoteVM
    : ComponentVMBase,
      ISelectable,
      IClosable,
      IDeletable<NoteVM>,
      ISavable<NoteVM>,
      IReconstructable
{
    private readonly Action<NoteVM>? _onClose;
    private readonly Func<NoteVM, Task>? _onDelete;
    private readonly Func<NoteVM, Task>? _onSave;
    private readonly Func<NoteVM, Task<bool>>? _confirmDelete;
    private readonly INotificationHub? _notificationHub;
    private readonly AsyncRelayCommand _innerDeleteCommand;
    private NoteModel _model;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Public hub accessor — see <see cref="NotebookVM.Hub"/> for rationale.</summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>Current note model. Equality-guarded setter — emits hub messages on change.</summary>
    public NoteModel Model
    {
        get => _model;
        set
        {
            if (EqualityComparer<NoteModel>.Default.Equals(_model, value)) return;
            var oldTitle = _model.Title;
            var oldStarred = _model.Starred;
            _model = value;
            NotifyPropertyChanged(nameof(Model));
            if (!string.Equals(oldTitle, value.Title, StringComparison.Ordinal))
            {
                NotifyPropertyChanged(nameof(Title));
            }
            if (oldStarred != value.Starred)
            {
                NotifyPropertyChanged(nameof(Starred));
            }
        }
    }

    /// <summary>Note id (proxy on <see cref="Model"/>).</summary>
    public string NoteId => _model.Id;

    /// <summary>Note title (proxy on <see cref="Model"/>).</summary>
    public string Title => _model.Title;

    /// <summary>Note body (proxy on <see cref="Model"/>).</summary>
    public string Body => _model.Body;

    /// <summary>Whether the note is starred (proxy on <see cref="Model"/>).</summary>
    public bool Starred => _model.Starred;

    /// <summary>Tag list (proxy on <see cref="Model"/>).</summary>
    public IReadOnlyList<string> Tags => _model.Tags;

    // ── IClosable / IDeletable / ISavable ──────────────────────────────────

    /// <inheritdoc cref="IClosable.CanClose"/>
    public bool CanClose() => Status == VMx.Lifecycle.ConstructionStatus.Constructed;

    /// <inheritdoc cref="IClosable.Close"/>
    public void Close() => _onClose?.Invoke(this);

    /// <inheritdoc cref="IDeletable{T}.CanDelete"/>
    public bool CanDelete(NoteVM item) => ReferenceEquals(item, this) && Status == VMx.Lifecycle.ConstructionStatus.Constructed;

    /// <inheritdoc cref="IDeletable{T}.Delete"/>
    public void Delete(NoteVM item)
    {
        if (!CanDelete(item)) return;
        _innerDeleteCommand.Execute(null);
    }

    /// <inheritdoc cref="ISavable{T}.CanSave"/>
    public bool CanSave(NoteVM item) => ReferenceEquals(item, this) && Status == VMx.Lifecycle.ConstructionStatus.Constructed;

    /// <inheritdoc cref="ISavable{T}.Save"/>
    public void Save(NoteVM item)
    {
        if (!CanSave(item)) return;
        SaveCommand.Execute(null);
    }

    /// <summary>
    /// Convenience command wrapper for <see cref="Close"/> — the view binds
    /// <see cref="CloseCommand"/> directly without going through capability
    /// dispatch.
    /// </summary>
    public ICommand CloseCommand { get; }

    /// <summary>Convenience command wrapper for <see cref="Save"/>.</summary>
    public ICommand SaveCommand { get; }

    /// <summary>Convenience command wrapper for <see cref="Delete"/>.</summary>
    public ICommand DeleteCommand { get; }

    private NoteVM(
        string name,
        string hint,
        NoteModel model,
        IMessageHub hub,
        IDispatcher dispatcher,
        Action<NoteVM>? onClose,
        Func<NoteVM, Task>? onDelete,
        Func<NoteVM, Task>? onSave,
        Func<NoteVM, Task<bool>>? confirmDelete,
        INotificationHub? notificationHub)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _model = model;
        _onClose = onClose;
        _onDelete = onDelete;
        _onSave = onSave;
        _confirmDelete = confirmDelete;
        _notificationHub = notificationHub;

        CloseCommand = RelayCommand.Builder()
            .Predicate(CanClose)
            .Task(Close)
            .Build();
        SaveCommand = AsyncRelayCommand.Builder()
            .Predicate(() => CanSave(this))
            .Task(_ => PerformSaveAsync(this))
            .Build();
        // Spec §5.2.8 / §6.2: Delete must be a ConfirmationDecoratorCommand.
        // When a confirm delegate is wired, wrap the inner delete command; if
        // accepted, post a "Note deleted" notification (if a hub is wired) and
        // invoke the host delete callback. Without a confirm delegate the
        // command stays plain — preserves tests that exercise the raw delete
        // path (e.g. NoteVMTests.DeleteCommand_invokes_OnDelete_callback).
        _innerDeleteCommand = AsyncRelayCommand.Builder()
            .Predicate(() => CanDelete(this))
            .Task(_ => PerformDeleteAsync(this))
            .Build();
        DeleteCommand = _confirmDelete is null
            ? _innerDeleteCommand
            : new ConfirmationDecoratorCommand(_innerDeleteCommand, () => _confirmDelete(this));
    }

    private async Task PerformDeleteAsync(NoteVM item)
    {
        if (!CanDelete(item)) return;
        if (_onDelete is not null)
            await _onDelete(item).ConfigureAwait(false);
        if (_notificationHub is not null)
        {
            _ = _notificationHub.Post(new Notification(
                NotificationType.Notification,
                $"Note deleted: “{item.Title}”"));
        }
    }

    private async Task PerformSaveAsync(NoteVM item)
    {
        if (!CanSave(item)) return;
        if (_onSave is not null)
            await _onSave(item).ConfigureAwait(false);
    }

    /// <inheritdoc/>
    protected override void OnDispose()
    {
        (CloseCommand as IDisposable)?.Dispose();
        (SaveCommand as IDisposable)?.Dispose();
        // Dispose the decorator (detaches its CanExecuteChanged relay) AND the
        // inner RelayCommand: ConfirmationDecoratorCommand.Dispose does not
        // dispose its inner command, so disposing only DeleteCommand would leak
        // the inner command in the decorated path. Idempotent when DeleteCommand
        // IS the inner (no confirm delegate wired).
        (DeleteCommand as IDisposable)?.Dispose();
        (_innerDeleteCommand as IDisposable)?.Dispose();
        base.OnDispose();
    }

    /// <summary>Returns a new empty builder for <see cref="NoteVM"/>.</summary>
    public static NoteVMBuilder Builder() => NoteVMBuilder.Empty;

    /// <summary>Immutable fluent builder (spec ch. 10).</summary>
    public sealed class NoteVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly NoteModel? _model;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly Action<NoteVM>? _onClose;
        private readonly Func<NoteVM, Task>? _onDelete;
        private readonly Func<NoteVM, Task>? _onSave;
        private readonly Func<NoteVM, Task<bool>>? _confirmDelete;
        private readonly INotificationHub? _notificationHub;

        internal static readonly NoteVMBuilder Empty = new();

        private NoteVMBuilder() { _hint = ""; }

        private NoteVMBuilder(
            string? name, string hint, NoteModel? model,
            IMessageHub? hub, IDispatcher? dispatcher,
            Action<NoteVM>? onClose, Func<NoteVM, Task>? onDelete, Func<NoteVM, Task>? onSave,
            Func<NoteVM, Task<bool>>? confirmDelete, INotificationHub? notificationHub)
        {
            _name = name; _hint = hint; _model = model;
            _hub = hub; _dispatcher = dispatcher;
            _onClose = onClose; _onDelete = onDelete; _onSave = onSave;
            _confirmDelete = confirmDelete; _notificationHub = notificationHub;
        }

        /// <summary>Sets the required Name.</summary>
        public NoteVMBuilder Name(string name) => new(name, _hint, _model, _hub, _dispatcher, _onClose, _onDelete, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets the optional Hint.</summary>
        public NoteVMBuilder Hint(string hint) => new(_name, hint, _model, _hub, _dispatcher, _onClose, _onDelete, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets the required note model.</summary>
        public NoteVMBuilder Model(NoteModel model) => new(_name, _hint, model, _hub, _dispatcher, _onClose, _onDelete, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets the required Services.</summary>
        public NoteVMBuilder Services(IMessageHub hub, IDispatcher dispatcher) => new(_name, _hint, _model, hub, dispatcher, _onClose, _onDelete, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets the optional close callback (NotesViewVM.Current = null).</summary>
        public NoteVMBuilder OnClose(Action<NoteVM> handler) => new(_name, _hint, _model, _hub, _dispatcher, handler, _onDelete, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets a synchronous delete callback.</summary>
        public NoteVMBuilder OnDelete(Action<NoteVM> handler) =>
            OnDelete(vm => { handler(vm); return Task.CompletedTask; });
        /// <summary>Sets an asynchronous delete callback (route to repo).</summary>
        public NoteVMBuilder OnDelete(Func<NoteVM, Task> handler) => new(_name, _hint, _model, _hub, _dispatcher, _onClose, handler, _onSave, _confirmDelete, _notificationHub);
        /// <summary>Sets a synchronous save callback.</summary>
        public NoteVMBuilder OnSave(Action<NoteVM> handler) =>
            OnSave(vm => { handler(vm); return Task.CompletedTask; });
        /// <summary>Sets an asynchronous save callback (route to repo).</summary>
        public NoteVMBuilder OnSave(Func<NoteVM, Task> handler) => new(_name, _hint, _model, _hub, _dispatcher, _onClose, _onDelete, handler, _confirmDelete, _notificationHub);
        /// <summary>
        /// Sets the optional confirm-delete delegate (spec §5.2.8 / §6.2).
        /// When set, <see cref="DeleteCommand"/> is wrapped in a
        /// <see cref="ConfirmationDecoratorCommand"/> that awaits this delegate
        /// before invoking the inner delete. Typically wires to
        /// <c>IDialogService.Confirm($"Delete “{n.Title}”?")</c> — matches the
        /// curly-quote emission used by NotesViewVM (and the Py/TS flavors).
        /// </summary>
        public NoteVMBuilder ConfirmDelete(Func<NoteVM, Task<bool>> confirm) => new(_name, _hint, _model, _hub, _dispatcher, _onClose, _onDelete, _onSave, confirm, _notificationHub);
        /// <summary>
        /// Sets the optional notification hub. When set, a successful delete
        /// (i.e. one that survived the confirm gate, if any) publishes a
        /// "Note deleted" notification.
        /// </summary>
        public NoteVMBuilder NotificationHub(INotificationHub hub) => new(_name, _hint, _model, _hub, _dispatcher, _onClose, _onDelete, _onSave, _confirmDelete, hub);

        /// <summary>Builds the VM after validating required fields.</summary>
        public NoteVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_model, "Model");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            return new NoteVM(_name!, _hint, _model!, _hub!, _dispatcher!, _onClose, _onDelete, _onSave, _confirmDelete, _notificationHub);
        }
    }
}

File: examples/typescript/react/notes-showcase/src/viewmodels/noteVM.ts
/**
 * NoteVM — leaf VM for a single note.
 *
 * Capabilities (scenario §6.2): ISelectable, IClosable, IDeletable<NoteVM>,
 * ISavable<NoteVM>, IReconstructable.
 *
 * `closeCommand` invokes a host-supplied close callback (the host wires it to
 * `NotesViewVM.current = null` so the form clears). This avoids back-
 * references from the leaf to its container. Same pattern as the C# / Python
 * flavors.
 */
import {
  ComponentVMBase,
  ConfirmationDecoratorCommand,
  ConstructionStatus,
  declareCapabilities,
  PropertyChangedMessage,
  RelayCommand,
  ViewModelType,
  type ICommand,
  type IDispatcher,
  type IMessageHub,
} from "vmx";
import {
  type INotificationHub,
  Notification,
  NotificationType,
} from "vmx/notifications";

import type { NoteModel } from "../models/noteModel.js";

const SENTINEL = Symbol("not-set");

type NoteHandler = (vm: NoteVM) => void;

export class NoteVM extends ComponentVMBase {
  readonly #onClose: NoteHandler | null;
  readonly #onDelete: NoteHandler | null;
  readonly #onSave: NoteHandler | null;
  readonly #confirmDelete: ((vm: NoteVM) => Promise<boolean>) | null;
  readonly #notificationHub: INotificationHub | null;
  readonly #closeCommand: RelayCommand;
  readonly #saveCommand: RelayCommand;
  readonly #innerDeleteCommand: RelayCommand;
  readonly #deleteCommand: ICommand;
  #model: NoteModel;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    model: NoteModel;
    onClose?: NoteHandler | null;
    onDelete?: NoteHandler | null;
    onSave?: NoteHandler | null;
    confirmDelete?: ((vm: NoteVM) => Promise<boolean>) | null;
    notificationHub?: INotificationHub | null;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#model = opts.model;
    this.#onClose = opts.onClose ?? null;
    this.#onDelete = opts.onDelete ?? null;
    this.#onSave = opts.onSave ?? null;
    this.#confirmDelete = opts.confirmDelete ?? null;
    this.#notificationHub = opts.notificationHub ?? null;
    declareCapabilities(
      this,
      "ISelectable",
      "IClosable",
      "IDeletable",
      "ISavable",
      "IReconstructable",
    );

    this.#closeCommand = RelayCommand.builder()
      .predicate(() => this.canClose())
      .task(() => this.close())
      .build();
    this.#saveCommand = RelayCommand.builder()
      .predicate(() => this.canSave(this))
      .task(() => this.save(this))
      .build();
    // Spec §5.2.8 / §6.2: when a confirm-delete delegate is wired, wrap the
    // delete in a ConfirmationDecoratorCommand. The inner command invokes
    // `#performDelete`, which posts a "Note deleted" notification (if a hub
    // is wired) and calls the host delete callback.
    this.#innerDeleteCommand = RelayCommand.builder()
      .predicate(() => this.canDelete(this))
      .task(() => this.#performDelete(this))
      .build();
    this.#deleteCommand =
      this.#confirmDelete !== null
        ? new ConfirmationDecoratorCommand(
            this.#innerDeleteCommand,
            () => this.#confirmDelete!(this),
          )
        : this.#innerDeleteCommand;
  }

  #performDelete(item: NoteVM): void {
    if (!this.canDelete(item)) return;
    if (this.#onDelete !== null) this.#onDelete(item);
    if (this.#notificationHub !== null) {
      void this.#notificationHub.post(
        new Notification(NotificationType.Notification, `Note deleted: “${item.title}”`),
      );
    }
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  get model(): NoteModel {
    return this.#model;
  }

  set model(value: NoteModel) {
    if (this.#model === value) return;
    const oldTitle = this.#model.title;
    const oldStarred = this.#model.starred;
    this.#model = value;
    this._hub.send(PropertyChangedMessage.create(this, this._name, "Model"));
    this._raisePropertyChanged("model");
    if (oldTitle !== value.title) {
      this._hub.send(PropertyChangedMessage.create(this, this._name, "Title"));
      this._raisePropertyChanged("title");
    }
    if (oldStarred !== value.starred) {
      this._hub.send(
        PropertyChangedMessage.create(this, this._name, "Starred"),
      );
      this._raisePropertyChanged("starred");
    }
  }

  /** Note id (proxy on `model`). */
  get noteId(): string {
    return this.#model.id;
  }

  get title(): string {
    return this.#model.title;
  }

  get body(): string {
    return this.#model.body;
  }

  get starred(): boolean {
    return this.#model.starred;
  }

  get tags(): readonly string[] {
    return this.#model.tags;
  }

  // ── IClosable ─────────────────────────────────────────────────────────────

  canClose(): boolean {
    return this.status === ConstructionStatus.Constructed;
  }

  close(): void {
    if (this.#onClose !== null) this.#onClose(this);
  }

  // ── IDeletable<NoteVM> ────────────────────────────────────────────────────

  canDelete(item: NoteVM): boolean {
    return item === this && this.status === ConstructionStatus.Constructed;
  }

  delete(item: NoteVM): void {
    if (!this.canDelete(item)) return;
    if (this.#onDelete !== null) this.#onDelete(item);
  }

  // ── ISavable<NoteVM> ──────────────────────────────────────────────────────

  canSave(item: NoteVM): boolean {
    return item === this && this.status === ConstructionStatus.Constructed;
  }

  save(item: NoteVM): void {
    if (!this.canSave(item)) return;
    if (this.#onSave !== null) this.#onSave(item);
  }

  get closeCommand(): ICommand {
    return this.#closeCommand;
  }

  get saveCommand(): ICommand {
    return this.#saveCommand;
  }

  get deleteCommand(): ICommand {
    return this.#deleteCommand;
  }

  protected override _onDispose(): void {
    this.#closeCommand.dispose();
    this.#saveCommand.dispose();
    // ConfirmationDecoratorCommand is not Disposable in VMx-TS; dispose the
    // raw inner RelayCommand explicitly to avoid leaking its CanExecute
    // subscriptions.
    this.#innerDeleteCommand.dispose();
    super._onDispose();
  }

  static builder(): NoteVMBuilder {
    return new NoteVMBuilder();
  }
}

export class NoteVMBuilder {
  #name: string | null = null;
  #hint = "";
  #model: NoteModel | typeof SENTINEL = SENTINEL;
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #onClose: NoteHandler | null = null;
  #onDelete: NoteHandler | null = null;
  #onSave: NoteHandler | null = null;
  #confirmDelete: ((vm: NoteVM) => Promise<boolean>) | null = null;
  #notificationHub: INotificationHub | null = null;

  constructor(from?: NoteVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#model = from.#model;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#onClose = from.#onClose;
      this.#onDelete = from.#onDelete;
      this.#onSave = from.#onSave;
      this.#confirmDelete = from.#confirmDelete;
      this.#notificationHub = from.#notificationHub;
    }
  }

  name(value: string): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#hint = value;
    return b;
  }

  model(value: NoteModel): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#model = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  onClose(fn: NoteHandler): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#onClose = fn;
    return b;
  }

  onDelete(fn: NoteHandler): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#onDelete = fn;
    return b;
  }

  onSave(fn: NoteHandler): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#onSave = fn;
    return b;
  }

  /** When set, `deleteCommand` is wrapped in a `ConfirmationDecoratorCommand`
   * calling this delegate. Typical host wiring:
   * `(vm) => dialogService.confirm(\`Delete "${vm.title}"?\`)`.
   */
  confirmDelete(fn: (vm: NoteVM) => Promise<boolean>): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#confirmDelete = fn;
    return b;
  }

  /** When set, a successful delete (post-confirm if any) posts a
   * "Note deleted" notification on the hub.
   */
  notificationHub(hub: INotificationHub): NoteVMBuilder {
    const b = new NoteVMBuilder(this);
    b.#notificationHub = hub;
    return b;
  }

  build(): NoteVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#model === SENTINEL) throw new Error("model is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services (hub + dispatcher) are required");
    return new NoteVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      model: this.#model,
      onClose: this.#onClose,
      onDelete: this.#onDelete,
      onSave: this.#onSave,
      confirmDelete: this.#confirmDelete,
      notificationHub: this.#notificationHub,
    });
  }
}
